GCC 6.0/6.x build problems

I get the next error compiling Opencv v2.4.12 in Arch linux https://aur.archlinux.org/packages/opencv2/:

---
[ 0%] Built target opencv_core_pch_dephelp
[ 0%] Generating precomp.hpp.gch/opencv_core_Release.gch
cc1plus: aviso: la opción de línea de órdenes ‘-Wmissing-prototypes’ es válida para C/ObjC pero no para C++
cc1plus: aviso: la opción de línea de órdenes ‘-Wstrict-prototypes’ es válida para C/ObjC pero no para C++
In file included from /usr/include/c++/6.1.1/bits/stl_algo.h:59:0,
from /usr/include/c++/6.1.1/algorithm:62,
from /tmp/packerbuild-1001/opencv2/opencv2/src/opencv-2.4.12.3/modules/core/include/opencv2/core/core.hpp:53,
**from /tmp/packerbuild-1001/opencv2/opencv2/src/opencv-2.4.12.3/modules/core/precomp.hpp:48:
/usr/include/c++/6.1.1/cstdlib:75:25: error fatal: stdlib.h: No existe el fichero o el directorio**
#include_next <stdlib.h>
^
compilación terminada.
modules/core/CMakeFiles/pch_Generate_opencv_core.dir/build.make:62: fallo en las instrucciones para el objetivo 'modules/core/precomp.hpp.gch/opencv_core_Release.gch'
make[2]: *** [modules/core/precomp.hpp.gch/opencv_core_Release.gch] Error 1
CMakeFiles/Makefile2:862: fallo en las instrucciones para el objetivo 'modules/core/CMakeFiles/pch_Generate_opencv_core.dir/all'
make[1]: *** [modules/core/CMakeFiles/pch_Generate_opencv_core.dir/all] Error 2
Makefile:160: fallo en las instrucciones para el objetivo 'all'
make: *** [all] Error 2

---

Here is the error message in english:
```
[  1%] Generating precomp.hpp
[  1%] Generating precomp.hpp.gch/opencv_core_Release.gch
In file included from /usr/include/c++/6.1.1/bits/stl_algo.h:59:0,
                 from /usr/include/c++/6.1.1/algorithm:62,
                 from /home/user/build/opencv2_2/src/opencv-2.4.12.3/modules/core/include/opencv2/core/core.hpp:53,
                 from /home/user/build/opencv2_2/src/opencv-2.4.12.3/modules/core/precomp.hpp:48:
/usr/include/c++/6.1.1/cstdlib:75:25: fatal error: stdlib.h: No such file or directory
 #include_next <stdlib.h>
                         ^
compilation terminated.
modules/core/CMakeFiles/pch_Generate_opencv_core.dir/build.make:62: recipe for target 'modules/core/precomp.hpp.gch/opencv_core_Release.gch' failed
make[2]: *** [modules/core/precomp.hpp.gch/opencv_core_Release.gch] Error 1
CMakeFiles/Makefile2:862: recipe for target 'modules/core/CMakeFiles/pch_Generate_opencv_core.dir/all' failed
make[1]: *** [modules/core/CMakeFiles/pch_Generate_opencv_core.dir/all] Error 2
Makefile:160: recipe for target 'all' failed
make: *** [all] Error 2
```

This is a behaviour change in GCC 6.1 for -isystem includes
See [https://gcc.gnu.org/bugzilla/show_bug.cgi?id=70129](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=70129) closed as WONTFIX
Sadly, I'm not that good with complex include dependencies, but I think this should be fixed to be compatible with upcoming gcc releases

@alalek Ahh, that makes sense

I tried the workaround but it doesn't change it, I get the exact same error message. It looks like the `-I` for system includes is not applied everywhere (A grep for `-isystem` still finds some occurences but not as many as without the workaround.

A (hacky) solution I came up with is this:
```
--- opencv-2.4.12.3/cmake/OpenCVPCHSupport.cmake
+++ opencv-2.4.12.3-patched/cmake/OpenCVPCHSupport.cmake
@@ -24,7 +24,7 @@
     ENDIF()
 
     SET(_PCH_include_prefix "-I")
-    SET(_PCH_isystem_prefix "-isystem")
+    SET(_PCH_isystem_prefix "-I")
     SET(_PCH_define_prefix "-D")
 
 ELSEIF(CMAKE_GENERATOR MATCHES "^Visual.*$")
```
Now it fails much later in the compile process with
```
[ 47%] Building CXX object modules/contrib/CMakeFiles/opencv_contrib.dir/src/chamfermatching.cpp.o
/home/user/build/opencv2_2/src/opencv-2.4.12.3/modules/contrib/src/chamfermatching.cpp: In member
 function ‘void cv::ChamferMatcher::Matching::computeDistanceTransform(cv::Mat&, cv::Mat&, cv::Mat&
, float, float, float)’:
/home/user/build/opencv2_2/src/opencv-2.4.12.3/modules/contrib/src/chamfermatching.cpp:969:30: er
ror: the compiler can assume that the address of ‘annotate_img’ will never be NULL [-Werror=address
]
             if (&annotate_img!=NULL) {
                              ^
/home/user/build/opencv2_2/src/opencv-2.4.12.3/modules/contrib/src/chamfermatching.cpp:1016:34: e
rror: the compiler can assume that the address of ‘annotate_img’ will never be NULL [-Werror=addres
s]
                 if (&annotate_img!=NULL) {
                                  ^
/home/user/build/opencv2_2/src/opencv-2.4.12.3/modules/contrib/src/chamfermatching.cpp: In member
 function ‘cv::ChamferMatcher::Match* cv::ChamferMatcher::Matching::localChamferDistance(cv::Point,
 cv::Mat&, cv::Mat&, cv::ChamferMatcher::Template*, float)’:
/home/user/build/opencv2_2/src/opencv-2.4.12.3/modules/contrib/src/chamfermatching.cpp:1111:25: e
rror: the compiler can assume that the address of ‘orientation_img’ will never be NULL [-Werror=add
ress]
     if (&orientation_img!=NULL) {
                         ^
/home/user/build/opencv2_2/src/opencv-2.4.12.3/modules/contrib/src/chamfermatching.cpp:1111:5: wa
rning: nonnull argument ‘orientation_img’ compared to NULL [-Wnonnull-compare]
     if (&orientation_img!=NULL) {
     ^~
/home/user/build/opencv2_2/src/opencv-2.4.12.3/modules/contrib/src/chamfermatching.cpp: In member
 function ‘void cv::ChamferMatcher::Matching::computeDistanceTransform(cv::Mat&, cv::Mat&, cv::Mat&
, float, float, float)’:
/home/user/build/opencv2_2/src/opencv-2.4.12.3/modules/contrib/src/chamfermatching.cpp:1016:17: w
arning: nonnull argument ‘annotate_img’ compared to NULL [-Wnonnull-compare]
                 if (&annotate_img!=NULL) {
                 ^~
/home/user/build/opencv2_2/src/opencv-2.4.12.3/modules/contrib/src/chamfermatching.cpp:969:13: wa
rning: nonnull argument ‘annotate_img’ compared to NULL [-Wnonnull-compare]
             if (&annotate_img!=NULL) {
             ^~
cc1plus: some warnings being treated as errors
modules/contrib/CMakeFiles/opencv_contrib.dir/build.make:158: recipe for target 'modules/contrib/CM
akeFiles/opencv_contrib.dir/src/chamfermatching.cpp.o' failed
make[2]: *** [modules/contrib/CMakeFiles/opencv_contrib.dir/src/chamfermatching.cpp.o] Error 1
CMakeFiles/Makefile2:2711: recipe for target 'modules/contrib/CMakeFiles/opencv_contrib.dir/all' fa
iled
make[1]: *** [modules/contrib/CMakeFiles/opencv_contrib.dir/all] Error 2
Makefile:160: recipe for target 'all' failed
make: *** [all] Error 2
```

--- a/cmake/OpenCVPCHSupport.cmake
+++ b/cmake/OpenCVPCHSupport.cmake
@@ -19,7 +19,7 @@ IF(CMAKE_COMPILER_IS_GNUCXX)
         ARGS ${CMAKE_CXX_COMPILER_ARG1} -dumpversion
         OUTPUT_VARIABLE gcc_compiler_version)
     #MESSAGE("GCC Version: ${gcc_compiler_version}")
-    IF(gcc_compiler_version VERSION_GREATER "4.2.-1")
+    IF(gcc_compiler_version VERSION_GREATER "4.2.-1" AND gcc_compiler_version VERSION_LESS "6.0.0")
         SET(PCHSupport_FOUND TRUE)
     ENDIF()
 

